torchgen/api/autograd.py

```diff
@@ -305,6 +306,164 @@ def dispatch_strategy(fn: NativeFunctionWithDifferentiabilityInfo) -> str:
+
+
+def is_reference_for_foreach(
+    f: NativeFunction,
+    function_schema: FunctionSchema,
+) -> bool:
+    return (
+        f.func.name.name.base.split("_foreach_")[-1] == function_schema.name.name.base
+        and not function_schema.name.name.inplace
+        and (
+            True
+            if len(f.func.arguments.post_self_positional) == 0
+            else all(
+                ref_arg.type in (arg.type, getattr(arg.type, "elem", None))
+                for arg, ref_arg in zip(
+                    f.func.arguments.flat_non_out,
+                    function_schema.arguments.flat_non_out,
+                )
+            )
+        )
+    )
+
+
+def gen_foreach_derivativeinfo(
+    foreach_function: NativeFunction,
+    differentiability_infos: Dict[FunctionSchema, Dict[str, DifferentiabilityInfo]],
+    functional_info_by_signature: Dict[
+        FunctionSchema, Dict[str, DifferentiabilityInfo]
+    ],
+) -> Optional[DifferentiabilityInfo]:
+    ref_diff_info: Optional[DifferentiabilityInfo] = None
+    for function_schema in functional_info_by_signature:
+        if not is_reference_for_foreach(foreach_function, function_schema):
+            continue
+        if function_schema in differentiability_infos:
+            ref_diff_info = differentiability_infos[function_schema]["Default"]
+        elif (
+            function_schema.signature(strip_default=True)
+            in functional_info_by_signature
+        ):
+            ref_diff_info = functional_info_by_signature[
+                function_schema.signature(strip_default=True)
+            ]["Default"]
+        else:
+            raise RuntimeError(
+                "Reference `DifferentiabilityInfo` for {} not found".format(
+                    foreach_function.func
+                )
+            )
+        if ref_diff_info is not None:
+            break
+    if ref_diff_info is None:
+        return None
+
+    refarg_indices, map_refarg2foreacharg, map_name2arg = {}, {}, {}
+    for i, (arg, ref_arg) in enumerate(
+        zip(
+            foreach_function.func.arguments.flat_non_out,
+            function_schema.arguments.flat_non_out,
+        )
+    ):
+        map_refarg2foreacharg[ref_arg.name] = arg.name
+        map_name2arg[arg.name] = arg
+        refarg_indices[ref_arg.name] = i
+
+    num_ref_derivatives = len(ref_diff_info.derivatives)
+    all_saved_inputs, all_saved_outputs, all_var_names = [], [], []
+    modified_derivative_formulas = []
+    for i, derivative in enumerate(ref_diff_info.derivatives):
+        modified_formula = derivative.formula.replace("grad", "grads[i]").replace(
+            "result", "result[i]"
+        )
+        saved_inputs, saved_outputs = [], []
+        with local.parametrize(
+            use_const_ref_for_mutable_tensors=foreach_function.use_const_ref_for_mutable_tensors,
+            use_ilistref_for_tensor_lists=foreach_function.part_of_structured_group,
+        ):
```

Comment:
Curious why we have this (I'm not too sure what it does). @bdhirsh do you know?

Reply:
The second flag is telling the codegen to spit out C++ kernels that use `ITensorListRef`, instead of `TensorList`, to represent the "list-of-tensors" jit type.

The context is this was an optimization that Yukio from Quansight worked on last year, that improves perf on ops that use tensorlists, because it makes it much cheaper to convert from boxed to unboxed representation (this has to do with the fact that the way that `IValue` internally stores lists of tensors is pretty inefficient).

Reply:
I vaguely remember that this path depends on this context just to call `cpp.argument_type` but I don't understand enough